### ISSUE TITLE
Is there any way we could make use of CallerArgumentExpressionAttribute?

### ISSUE THREAD

**Issue by Dreamescaper** [nunit/nunit#3936]
C#10 has a nice addition of [CallerArgumentExpressionAttribute](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/caller-argument-expression.md), which we could use to print `actual` argument expression in error messages.

I can see two problems though:
- We need to add optional `string` parameter to methods. But we already have overloads with `message` string parameter.
- We have methods with `params` parameter. And we can't combine `params` and optional parameters in the same method.

Is there any way we could make use of this attribute?

**Comment by manfred-brands** [comment]
We cannot do this with breaking binary compatibility, but the attribute is only effective at compile time anyhow.
For source code we have currently two methods for each test, e.g.:
```
Assert.True(bool condition)
Assert.True(bool condition, string? message, param object?[]? args)
```
We can replace the first with:
```
Assert.True(bool condition, [CallerArgumentExpression("condition")] string? message = null)
```

It will not be confused with the 2nd. Similar like `Console.WriteLine` has different overloads.

**Comment by Dreamescaper** [comment]
Problem is that existing code like that:
```csharp
Assert.True(myCondition, "MyCondition should certainly be true!");
```
will pick up the first method now.
> Assert.True(bool condition, [CallerArgumentExpression("condition")] string? conditionExpression = null)

**Comment by manfred-brands** [comment]
Which is correcy. The CallerArgumentExpression only evaluates to the condition string when not specified:

```
Assert.True(args.Length == 0);
Assert.True(args.Length == 0, "args");
Assert.True(args.Length == 0, "{0} != 0", args.Length);
```

Gives:
```
Expected 'true' for 'args.Length == 0'
Expected 'true' for 'args'
Expected 'true' for '1 != 0'
```

**Comment by Dreamescaper** [comment]
Problem is that existing method usages rarely provide `actual` expression, but usually some failure description.
So for my example
```csharp
Assert.True(myCondition, "MyCondition should be true when the sky is blue!");
```

it will print 
> Expected 'true' for 'MyCondition should be true when the sky is blue!'

which is probably not what's expected.

**Comment by manfred-brands** [comment]
Yes, this only helps where there was no message given previously.

**Comment by manfred-brands** [comment]
This one helps with your situation:
```
public static void True(bool condition, string? userMessage = null, [CallerArgumentExpression("condition")] string? conditionMessage = null)
```

This however will conflict in one case with the `params` overload if a user passes a string argument as params, e.g.: 
`Assert.True(variable, "Expected: '{0}' to be true.", nameof(variable));`

Potentially this can be detected by scanning for '{0' in the string.

**Comment by jamers99** [comment]
This would be a really great feature to have. Is there any option to do this ourselves?
